src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs

```diff
@@ -1315,6 +1324,500 @@ enum H : C { }
+                // (5,27): error CS8145: Auto-implemented properties cannot return by reference
+                //     data ref readonly int P2;
+                Diagnostic(ErrorCode.ERR_AutoPropertyCannotBeRefReturning, "P2").WithArguments("C.P2").WithLocation(5, 27),
+                // (6,10): error CS1519: Invalid token 'out' in class, struct, or interface member declaration
+                //     data out int P3;
+                Diagnostic(ErrorCode.ERR_InvalidMemberDecl, "out").WithArguments("out").WithLocation(6, 10),
+                // (6,10): error CS1519: Invalid token 'out' in class, struct, or interface member declaration
+                //     data out int P3;
+                Diagnostic(ErrorCode.ERR_InvalidMemberDecl, "out").WithArguments("out").WithLocation(6, 10),
+                // (6,18): warning CS0169: The field 'C.P3' is never used
+                //     data out int P3;
+                Diagnostic(ErrorCode.WRN_UnreferencedField, "P3").WithArguments("C.P3").WithLocation(6, 18)
+            );
+        }
+
+        [Fact]
+        public void DataPropertiesInterface()
+        {
+            var src = @"
+interface I
+{
+    data int P1;
+    data int P2 = 0;
+    static data int P3;
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (5,14): error CS8053: Instance properties in interfaces cannot have initializers.
+                //     data int P2 = 0;
+                Diagnostic(ErrorCode.ERR_InstancePropertyInitializerInInterface, "P2").WithArguments("I.P2").WithLocation(5, 14),
+                // (6,21): error CS0106: The modifier 'static' is not valid for this item
+                //     static data int P3;
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "P3").WithArguments("static").WithLocation(6, 21)
+            );
+
+            var p1 = comp.GlobalNamespace.GetTypeMember("I").GetMember<PropertySymbol>("P1");
+            Assert.True(p1.IsAbstract);
+            Assert.True(p1.GetMethod.IsAbstract);
+            Assert.True(p1.SetMethod.IsAbstract);
+            Assert.True(p1.SetMethod.IsInitOnly);
+        }
+
+        [Fact]
+        public void DataPropertiesOverride()
+        {
+            var src = @"
+abstract class C1
+{
+    data int P1;
+    virtual data int P2;
+    abstract data int P3;
+}
+abstract class C2 : C1
+{
+    data int P1; // warn 1
+    data int P2; // warn 2
+    data int P3; // error
+}
+class C3 : C1
+{
+    new data int P1;
+    new data int P2;
+    override data int P3;
+}
+class C4 : C1
+{
+    public new int P1 { get; init; }
+    public new virtual int P2 { get; }
+    public override int P3 { get; init; }
+}
+class C5 : C4
+{
+    new data int P1;
+    override data int P2; // error
+    override data int P3;
+}";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
```

Comment:
>comp.VerifyDiagnostics( [](start = 12, length = 23)

We should have tests for abstract data properties, i.e. verify that they are indeed abstract. Should also test success scenarios for overriding, including executing and observing expected runtime behavior.

Reply:
Also, tests with sealed. 

---
In reply to: [447692972](https://github.com/dotnet/roslyn/pull/44883#discussion_r447692972) [](ancestors = 447692972)

Reply:
>  don't think special execution for data properties tests is particularly interesting (since they're just syntax sugar), unless there's something different in the lowering for data properties.

Well, there is a dedicated code that processes modifiers, etc.. In general we don't assume that things would work just because we  expect them to. In fact, we almost always expect them to work, but that isn't always the reality. I am not saying we have to test full test matrix as for regular properties, but to have at least one success case verifying runtime behavior for each distinct flavor, is a must, in my opinion. Especially that overriding is such an interesting case, takes unique code paths, etc.

---
In reply to: [448561647](https://github.com/dotnet/roslyn/pull/44883#discussion_r448561647) [](ancestors = 448561647)